### PR TITLE
Adds implementation of entmax for TabNet, replaces sparsemax implementation in torch_utils.py 

### PR DESCRIPTION
**Pull request by dantreiman**
Adds implementation of entmax for TabNet.

Entmax is a family of transformations which generalizes softmax and sparsemax, introduced in  [Sparse Sequence-to-Sequence Models](https://arxiv.org/abs/1905.05702).

This PR makes use of the authors' implementation [https://github.com/deep-spin/entmax](https://github.com/deep-spin/entmax).

Adds two keys to tabnet combiner config: `entmax_mode: str {'sparsemax', 'entmax15', 'constant', 'adaptive'}` and `entmax_alpha: float [1.0, 2.0], default=1.5`.

- 'sparsemax' is the default, and corresponds to entmax with alpha 2.0.
- 'entmax15' is an optimized implementation of entmax with alpha 1.5
- 'constant' uses entmax with a constant alpha (entmax_alpha)
- 'adaptive' makes entmax_alpha a trainable parameter, attempts to learn the optimal sparsity by gradient descent.

On some datasets, using entmax will give small performance improvement over sparsemax.  For example, training `train_higgs_best.yaml` for 100 epochs:

Sparsemax
```
╒═════════╤════════╤═══════════╤════════════╕
│ label   │   loss │   roc_auc │   accuracy │
╞═════════╪════════╪═══════════╪════════════╡
│ train   │ 0.4416 │    0.8733 │     0.7875 │
├─────────┼────────┼───────────┼────────────┤
│ test    │ 0.4521 │    0.8669 │     0.7811 │
╘═════════╧════════╧═══════════╧════════════╛
╒════════════╤════════╕
│ combined   │   loss │
╞════════════╪════════╡
│ train      │ 0.4424 │
├────────────┼────────┤
│ test       │ 0.4523 │
╘════════════╧════════╛
```

Entmax 1.5
```
╒═════════╤════════╤═══════════╤════════════╕
│ label   │   loss │   roc_auc │   accuracy │
╞═════════╪════════╪═══════════╪════════════╡
│ train   │ 0.4263 │    0.8825 │     0.7965 │
├─────────┼────────┼───────────┼────────────┤
│ test    │ 0.4515 │    0.8673 │     0.7818 │
╘═════════╧════════╧═══════════╧════════════╛
╒════════════╤════════╕
│ combined   │   loss │
╞════════════╪════════╡
│ train      │ 0.4272 │
├────────────┼────────┤
│ test       │ 0.4518 │
╘════════════╧════════╛
```

Entmax adaptive
```
╒═════════╤════════╤═══════════╤════════════╕
│ label   │   loss │   roc_auc │   accuracy │
╞═════════╪════════╪═══════════╪════════════╡
│ train   │ 0.4128 │    0.8907 │     0.8049 │
├─────────┼────────┼───────────┼────────────┤
│ test    │ 0.4583 │    0.8647 │     0.7804 │
╘═════════╧════════╧═══════════╧════════════╛
╒════════════╤════════╕
│ combined   │   loss │
╞════════════╪════════╡
│ train      │ 0.4128 │
├────────────┼────────┤
│ test       │ 0.4585 │
╘════════════╧════════╛
```

### CODE REVIEW

**Review by tgaddair — APPROVED** [review]
Looks great!
